### ISSUE TITLE
DIABLO - accept a list of user defined folds in `perf()` and `perf.assess()`

### ISSUE THREAD

**Issue by dereklei12** [mixOmicsTeam/mixOmics#400]
**Is your feature request related to a problem? Please describe.**

Many studies have samples that belong to groups. For example several samples from the same subject, the same plot, the same site, the same tank, or the same batch. In these studies the real question is whether the model works on a new group. To answer that, all samples of one group must be in the same test fold.

Right now `perf()` and `perf.assess()` for DIABLO models always build the folds internally. The `folds` argument must be a single number. Samples are then split at random (stratified by class). Samples from the same group end up in both the training set and the test set. The model is tested on groups it has already seen, so the error rate looks better than it really is.

There is no way to give your own folds. The documentation of `perf()` says that `folds` "can be supplied as a list of vectors containing the indexes defining each fold as produced by split". This already works for pls and spls models. For DIABLO models the same list gives an error. So the behaviour is not consistent across models and the documentation is wrong for DIABLO.

**Reproducible example:**

```r
library(mixOmics)
data(breast.TCGA)
X <- list(mRNA  = breast.TCGA$data.train$mrna,
          miRNA = breast.TCGA$data.train$mirna)
Y <- breast.TCGA$data.train$subtype
n <- length(Y)

# pretend samples come from 5 groups (for example 5 sites or 5 batches)
group <- rep(1:5, length.out = n)
my.folds <- split(seq_len(n), group)

diablo <- block.splsda(X, Y, ncomp = 2,
                       keepX = list(mRNA = c(10, 10), miRNA = c(10, 10)))

# integer folds work
res1 <- perf(diablo, validation = "Mfold", folds = 5, nrepeat = 1, seed = 1)

# the same call with a list of folds fails
res2 <- perf(diablo, validation = "Mfold", folds = my.folds, nrepeat = 1)
#> Error : BiocParallel errors
#>   1 remote errors, element index: 1
#>   0 unevaluated and other errors
#>   first remote error:
#> Error in repeat_cv_perf.diablo(nrep): Invalid number of folds.

# the same list works for an spls model
spls.model <- spls(X$mRNA, X$miRNA, ncomp = 2, keepX = c(10, 10))
res3 <- perf(spls.model, validation = "Mfold", folds = my.folds, nrepeat = 1)
# runs fine, 5 folds used
```

Tested with R 4.6.1 and mixOmics 6.36.0. The code path is the same in the current master (6.37.0).

Two users asked for this on the forum, in November 2024 and again in January 2026:
https://mixomics-users.discourse.group/t/group-wise-cross-validation/1719

**Describe the solution you'd like**

This issue only covers DIABLO models, that is `perf()` and `perf.assess()` on a `block.splsda` object. The same change for other models (plsda, splsda, tune functions) will be opened as separate issues so that each pull request stays small.

1. Accept a list in `folds`. When `validation = "Mfold"` and `folds` is a list, use each element of the list as the test set of one fold. The training set is all other samples. Skip the internal call to `stratified.subsampling()`. This needs to change in two places that share the same code: `repeat_cv_perf.diablo()` in `R/perf.diablo.R` (around line 114) and `repeat_cv_perf.assess.diablo()` in `R/perf.assess.diablo.R` (around line 114).

2. Reuse the existing checks. `perf.mixo_pls()` in `R/perf.R` (lines 428 to 441) already checks a list of folds: at least 2 folds, at most n folds, the total number of indices equals n, and no index appears twice. Move these checks into one internal helper, for example `.check_folds()`, and call it from both the pls and the DIABLO methods. Add one more check that every index is a whole number between 1 and n. Add a warning when one class is missing from a training set, in the same spirit as the current warning about classes not represented in a fold.

3. Handle `nrepeat`. User defined folds are fixed, so repeating the cross validation gives the same result every time. When `folds` is a list and `nrepeat > 1`, give a warning and set `nrepeat` to 1. This is the same pattern already used for `validation = "loo"` in `.check_cv_args()`.

4. Update the documentation. In `perf()` and `perf.assess()`, the `@param folds` entry should say that `folds` is either a positive integer (number of folds) or a list of integer vectors where each element gives the row indices of one test fold. It should also say which models accept a list: pls, spls and DIABLO. The `perf.assess()` entry currently says "numeric" only, which is wrong even for pls.

5. Add tests in `tests/testthat/test-perf.and.perf.assess.diablo.R`. A list of folds runs without error for both `perf()` and `perf.assess()`. The number of folds in the output equals the length of the list. Every sample is predicted exactly once. A list with a duplicated index, a missing sample, or an index larger than n gives a clear error. A list combined with `nrepeat > 1` gives a warning and runs once.

Expected usage after the change:

```r
my.folds <- split(seq_len(nrow(X$mRNA)), group)
perf(diablo, validation = "Mfold", folds = my.folds)
perf.assess(diablo, validation = "Mfold", folds = my.folds)
```

**Describe alternatives you've considered**

1. The workaround suggested on the forum: copy `repeat_cv_perf.diablo()` from the package source, edit it, and put it back into the namespace with `environment(...) <- asNamespace("mixOmics")`. This works but needs knowledge of the package internals and has to be redone after every release.

2. The `multilevel` argument. For splsda and spls models, `multilevel` keeps all samples of one subject in the same fold during cross validation. But it also forces the within subject decomposition, and `block.splsda()` does not have a `multilevel` argument at all. So this is not an option for DIABLO.

3. Writing your own loop over folds with `block.splsda()` and `predict()`. This loses the outputs that `perf()` produces, such as the majority vote and weighted vote error rates, the feature stability, and the choice of the number of components.

**Additional context**

Where the error comes from: in `R/perf.diablo.R` the check is `if (is.null(folds) || !is.numeric(folds) || folds < 2 || folds > n)`. A list is not numeric, so the function stops with "Invalid number of folds" before it looks at the content of the list.

Out of scope for this issue, to be opened separately:

1. `perf()` and `perf.assess()` for plsda and splsda, and `tune.splsda()` and `tune.spls()`. These all go through `MCVfold.spls()` in `R/MCV.spls.R`, so they can be done together in one change.
2. `tune.block.splsda()` and `tune.block.plsda()`, which go through `MCVfold.block.splsda()`.
3. A standard deviation across folds when `nrepeat = 1`. `perf()` for DIABLO computes one pooled error rate per repeat from all folds together, so there is no per fold error rate today. Adding one is a new output and needs its own discussion.

The same feature exists in other packages: `caret::trainControl(indexOut = ...)`, `rsample::group_vfold_cv()`, and the group column in `mlr3` resampling.
